Source/Core/Core/IOS/DolphinDevice.cpp

```diff
@@ -0,0 +1,148 @@
+#include "Core/HW/Memmap.h"
+#include "Core/IOS/DolphinDevice.h"
+
+namespace IOS::HLE::Device
+{
+namespace
+{
+enum
+{
+  IOCTL_DOLPHIN_GET_SYSTEM_TIME = 0x01,
+  IOCTL_DOLPHIN_GET_VERSION = 0x02,
+  IOCTL_DOLPHIN_GET_SPEED_LIMIT = 0x03,
+  IOCTL_DOLPHIN_SET_SPEED_LIMIT = 0x04,
+  IOCTL_DOLPHIN_GET_CPU_SPEED = 0x05,
+
+};
+
+IPCCommandResult GetSystemTime(const IOCtlVRequest& request)
+{
+  if (!request.HasNumberOfValidVectors(0, 1))
+  {
+    return DolphinDevice::GetDefaultReply(IPC_EINVAL);
+  }
+
+  if (request.io_vectors[0].size != 4)
+  {
+    return DolphinDevice::GetDefaultReply(IPC_EINVAL);
+  }
+
+  const u32 milliseconds = Common::Timer::GetTimeMs();
+
+  Memory::Write_U32(milliseconds, request.io_vectors[0].address);
+  return DolphinDevice::GetDefaultReply(IPC_SUCCESS);
+}
+
+IPCCommandResult GetVersion(const IOCtlVRequest& request)
+{
+  if (!request.HasNumberOfValidVectors(0, 1))
+  {
+    return DolphinDevice::GetDefaultReply(IPC_EINVAL);
+  }
+
+  const auto length = std::min(size_t(request.io_vectors[0].size), std::strlen(SCM_DESC_STR));
+
+  Memory::Memset(request.io_vectors[0].address, 0, request.io_vectors[0].size);
+  Memory::CopyToEmu(request.io_vectors[0].address, SCM_DESC_STR, length);
+
+  return DolphinDevice::GetDefaultReply(IPC_SUCCESS);
+}
+
+IPCCommandResult GetCPUSpeed(const IOCtlVRequest& request)
+{
+  if (!request.HasNumberOfValidVectors(0, 1))
+  {
+    return DolphinDevice::GetDefaultReply(IPC_EINVAL);
+  }
+
+  if (request.io_vectors[0].size != 4)
+  {
+    return DolphinDevice::GetDefaultReply(IPC_EINVAL);
+  }
+
+  const SConfig& config = SConfig::GetInstance();
+  const float oc = config.m_OCEnable ? config.m_OCFactor : 1.0f;
+
+  const u32 core_clock = u32(float(SystemTimers::GetTicksPerSecond()) * oc);
+
+  Memory::Write_U32(core_clock, request.io_vectors[0].address);
+
+  return DolphinDevice::GetDefaultReply(IPC_SUCCESS);
+}
+
+IPCCommandResult GetSpeedLimit(const IOCtlVRequest& request)
+{
+  if (request.HasNumberOfValidVectors(0, 1))
+  {
+    // get current speed limit
+    if (request.io_vectors[0].size != 4)
+    {
+      return DolphinDevice::GetDefaultReply(IPC_EINVAL);
+    }
+
+    const SConfig& config = SConfig::GetInstance();
+    const u32 speed_percent = config.m_EmulationSpeed * 100;
+    Memory::Write_U32(speed_percent, request.io_vectors[0].address);
+
+    return DolphinDevice::GetDefaultReply(IPC_SUCCESS);
+  }
+  return DolphinDevice::GetDefaultReply(IPC_EINVAL);
+}
+
+IPCCommandResult SetSpeedLimit(const IOCtlVRequest& request)
+{
+  if (request.HasNumberOfValidVectors(1, 0))
+  {
+    // set current speed limit
+    if (request.in_vectors[0].size != 4)
+    {
+      return DolphinDevice::GetDefaultReply(IPC_EINVAL);
+    }
+
+    const float speed = float(Memory::Read_U32(request.in_vectors[0].address)) / 100.0f;
+    SConfig::GetInstance().m_EmulationSpeed = speed;
+    BootManager::SetEmulationSpeedReset(true);
+  }
+
+  return DolphinDevice::GetDefaultReply(IPC_EINVAL);
```

Comment:
Should this not be returning IPC_SUCCESS? Currently both the successful path and the error path both return the same error code, which is likely not intentional.